homeassistant/components/greeneye_monitor/sensor.py

```diff
@@ -118,13 +117,18 @@ def __init__(
         """Construct the entity."""
         self._monitor_serial_number = monitor_serial_number
         self._attr_name = name
-        self._sensor: T | None = None
+        self._sensor: UnderlyingSensorType | None = None
         self._sensor_type = sensor_type
         self._number = number
         self._attr_unique_id = (
             f"{self._monitor_serial_number}-{self._sensor_type}-{self._number}"
         )
 
+    def _set_sensor(self, sensor: UnderlyingSensorType) -> None:
+        self._sensor = sensor
```

Comment:
Ah, in the other PR when i suggested a change I didn't realize this parent class is still holding on to _sensor for `async_will_remove_from_hass`. This means that there is still a little bit of back and forth between the parent class in `_try_connect_to_monitor`, down into the child class in `_connect_to_monitor` then back up to the parent in `_set_sensor` which is unfortunate.

I think if you go with a structure like this then the parent does not need to retain sensor:
```
class CurrentSensor(GEMSensor):
    ...
    def add_listener(self, monitor: greeneye.monitor.Monitor) -> Callable[...]
        sensor = monitor.channels[self._number - 1]
        sensor.add_listener(self.async_write_ha_state)
        return partial(sensor.remove_listener, self.async_write_ha_state)
```

Then the parent can do:
```
def _try_connect_to_monitor(self, monitors: Monitors) -> bool:
    ...
    self._unsub = self.add_listener(monitor)
```
and `async_will_remove_from_hass` can invoke `self._unsub()` if set.

Reply:
Ah, OK, you were after the calls passing back and forth, not just the generics. I think I can duplicate less code (and avoid the partial functions) by storing Monitor and having each subclass just implement a getter for the sensor.